### PR TITLE
Phone lookup

### PR DESCRIPTION
**Pull request by alxdarksage**
Introduces an AccountId object which is used by AccountDao to retrieve accounts. It contains a study ID and one of: an ID, email or phone number. You still can't sign in using a phone #, that'll come next.

Once you start tugging at this thread you realize we have many, many methods that take a studyIdentifier, and then a userId, email, or eventually a phone number. So I just stopped at a place where everything compiles, calling AccountDao and the related AccountWorflowService. This is already a lot of files.

The only functional change was one place where we were returning a 404 if someone requested a password reset email. We should not do this, it allows you to determine which emails are in the study. It also fit better with a refactor to use AccountId.

The AccountId will throw an exception if you use a getter on the object for a field that has not been initialized. This should make it easier to verify that it's not being constructed with one identifier, then being called to retrieve a different identifier (e.g. made with a userId, then called later to retrieve an email address). The only place this isn't relevant is in the private getHibernateAccount() method where we'll query the DB based on what we find in AccountId. There's an accessor on AccountId to get a version that doesn't have these precondition checks.

### CODE REVIEW

File: app/org/sagebionetworks/bridge/hibernate/HibernateAccountDao.java [hunk 1]
```diff
@@ -403,27 +394,38 @@ private String hashCredential(PasswordAlgorithm algorithm, String type, String v
...
 
     /** {@inheritDoc} */
     @Override
-    public void deleteAccount(Study study, String id) {
-        hibernateHelper.deleteById(HibernateAccount.class, id);
+    public void deleteAccount(AccountId accountId) {
+        HibernateAccount hibernateAccount = getHibernateAccount(accountId);
+        if (hibernateAccount != null) {
+            hibernateHelper.deleteById(HibernateAccount.class, hibernateAccount.getId());    
+        }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/hibernate/HibernateAccountDao.java]: This is functionally identical, but in the new code, in the string ID case, we'll query the DB for an account, and then only use the string ID (which we already have). Is it worth adding a hasId() to AccountId and optimizing on the case where we only need the string ID?

File: test/org/sagebionetworks/bridge/models/accounts/AccountIdTest.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                .equals(AccountId.forEmail(TEST_STUDY_IDENTIFIER, "email")));
+    }
+    
+    @Test
+    public void factoryMethodsWork() {
+        String number = PHONE.getNumber();
+        assertEquals("one", AccountId.forId(TEST_STUDY_IDENTIFIER, "one").getId());
+        assertEquals("one", AccountId.forEmail(TEST_STUDY_IDENTIFIER, "one").getEmail());
+        assertEquals(number, AccountId.forPhone(TEST_STUDY_IDENTIFIER, PHONE).getPhone().getNumber());
+    }
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/models/accounts/AccountIdTest.java:38]: Is it worth adding tests that get() throws if the value is null?

File: test/org/sagebionetworks/bridge/models/accounts/SignInTest.java [hunk 3]
```diff
@@ -89,5 +89,23 @@ public void acceptsUsernameAsEmail() throws Exception {
...
+    }
+    
+    @Test
+    public void signInAccountIdWithPhone() {
+        SignIn signIn = new SignIn.Builder().withStudy(TestConstants.TEST_STUDY_IDENTIFIER)
+                .withPhone(TestConstants.PHONE).withPassword("password").build();
+        AccountId accountId = signIn.getAccountId();
+        assertEquals(TestConstants.TEST_STUDY_IDENTIFIER, accountId.getStudyId());
+        assertEquals(TestConstants.PHONE.getNumber(), accountId.getPhone().getNumber());
+    }
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/models/accounts/SignInTest.java:111]: Is it worth adding a case for getAccountId() with neither email nor phone?

File: app/org/sagebionetworks/bridge/services/AccountWorkflowService.java [hunk 4]
```diff
@@ -168,29 +167,29 @@ public void notifyAccountExists(Study study, Email email) {
...
-        checkNotNull(email);
+    public void requestResetPassword(AccountId accountId) {
+        checkNotNull(accountId);
         
-        Account account = accountDao.getAccountWithEmail(study, email.getEmail());
+        Account account = accountDao.getAccount(accountId);
         if (account != null) {
-            sendPasswordResetRelatedEmail(study, email, study.getResetPasswordTemplate());    
+            Study study = studyService.getStudy(account.getStudyIdentifier());
+            sendPasswordResetRelatedEmail(study, account.getEmail(), study.getResetPasswordTemplate());
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/services/AccountWorkflowService.java]: It looks like in the old code, we already take in study. Would it be better to keep that, instead of calling studyService to get the study again?

File: test/org/sagebionetworks/bridge/services/ParticipantServiceTest.java [hunk 5]
```diff
@@ -724,16 +730,17 @@ public void requestResetPassword() {
         
         participantService.requestResetPassword(STUDY, ID);
         
-        verify(accountDao).requestResetPassword(eq(STUDY), emailCaptor.capture());
+        verify(accountDao).requestResetPassword(accountIdCaptor.capture());
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/services/ParticipantServiceTest.java]: Nitpick: Would eq(ACCOUNT_ID) be better than use a capture?

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]
Mostly good, a few small comments.

File: app/org/sagebionetworks/bridge/hibernate/HibernateAccountDao.java (hunk 1 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/hibernate/HibernateAccountDao.java]: Changed to check for ID and only query if it has something else

File: test/org/sagebionetworks/bridge/models/accounts/AccountIdTest.java (hunk 2 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/models/accounts/AccountIdTest.java:38]: Added. Also changed the guard because it was throwing IllegalArgumentException on a getter with no parameters, which seems weird. Now it throws an NPE.

File: app/org/sagebionetworks/bridge/services/AccountWorkflowService.java (hunk 4 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/services/AccountWorkflowService.java]: Okay, carried study over from services. Added a precondition check that study ID and account ID are the same.

File: test/org/sagebionetworks/bridge/models/accounts/SignInTest.java (hunk 3 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/models/accounts/SignInTest.java:111]: Added. Verified we validate before calling getAccountId() in all pathways since this will generate a server error.

File: test/org/sagebionetworks/bridge/services/ParticipantServiceTest.java (hunk 5 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/services/ParticipantServiceTest.java]: Yes, it's better, changed.

File: test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java [hunk 6]
```diff
@@ -850,13 +851,72 @@ public void getByEmailNotFound() {
...
+        
+        verify(mockHibernateHelper).deleteById(HibernateAccount.class, ACCOUNT_ID);
+    }
+    
+    @Test
+    public void deleteWithId() throws Exception {
+        // Directly deletes with the ID it has
+        dao.deleteAccount(ACCOUNT_ID_WITH_ID);
+        
         verify(mockHibernateHelper).deleteById(HibernateAccount.class, ACCOUNT_ID);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java:920]: Is it worth adding a check that we *don't* re-query Hibernate for the account?

File: test/org/sagebionetworks/bridge/models/accounts/SignInTest.java [hunk 7]
```diff
@@ -89,5 +89,36 @@ public void acceptsUsernameAsEmail() throws Exception {
...
+    
+    @Test
+    public void signInAccountIncomplete() {
+        SignIn signIn = new SignIn.Builder().withStudy(TestConstants.TEST_STUDY_IDENTIFIER)
+                .withPassword("password").build();
+        // SignIn should be validated to hold either email or phone before we 
+        // retrieve accountId 
+        try {
+            signIn.getAccountId();    
+        } catch(IllegalArgumentException e) {
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/models/accounts/SignInTest.java]: Need a fail() here. Otherwise, this test passes if getAccountId() doesn't throw.

**Review by DwayneJengSage — APPROVED** [review]
Just two more small comments about test. Otherwise, looks good.

File: test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java (hunk 6 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java:920]: Added

File: test/org/sagebionetworks/bridge/models/accounts/SignInTest.java (hunk 7 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/models/accounts/SignInTest.java]: good catch, added
